File: src/22.括号生成.ts
/*
 * @lc app=leetcode.cn id=22 lang=typescript
 *
 * [22] 括号生成
 */

// @lc code=start
function generateParenthesis(n: number): string[] {
    let res: string[] = []

    // baoli
    const backtracks = (open: number, close: number, temp: string, max: number, res: string[]) => {
        if(temp.length >= max * 2) {
            res.push(temp)
            return
        }

        if(open < max) {
            backtracks(open + 1, close, temp + '(', max, res)
        }

        if( close < open) {
            backtracks(open, close + 1, temp + ')', max, res)
        }
    }

    backtracks(0, 0, '', n, res)

    return res
};
// @lc code=end

File: src/22.括号生成.go
/*
 * @lc app=leetcode.cn id=22 lang=golang
 *
 * [22] 括号生成
 */
package leetcode

// @lc code=start
func generateParenthesis(n int) (ans []string) {
	var (
		dfs func(lcnt, rcnt int, s string)
	)

	dfs = func(lcnt, rcnt int, s string) {
		if lcnt > n || rcnt > n || rcnt>lcnt {
			return
		}

		if lcnt == n && rcnt == n {
			ans = append(ans, s)
			return
		}

		dfs(lcnt+1,rcnt,s+"(")
		dfs(lcnt,rcnt+1,s+")")
	}

	dfs(0,0,"")

	return
}
// @lc code=end

